### ISSUE TITLE
transcribe() outputting different results than validate()

### ISSUE THREAD

**Issue by carlosfranzreb** [NVIDIA/NeMo#5289]
Hello,

I have fine-tuned the German Conformer Transducer on a subset of Mozilla Common Voice. The WER on the validation set decreases throughout the 20 training epochs. However, when I load the checkpoint and use the `transcribe` method, the model outputs weird predictions containing `??`. For example: 

```
' ⁇ un müssen meines  ⁇ rachtens konkrete und erkennbare  ⁇ aßnahmen ergriffen werden ⁇ '
```

This does not occur when I use the pre-trained model offered by Nvidia. Only when I use the one I have fine-tuned myself. I have run another validation loop with the same checkpoint, and the WER is lower than for the model that is not fine-tuned. 

I assume that there must be a difference between these two methods (`validate` and `transcribe`) that makes one work but not the other. Do you have any clue as to why this is?

**Comment by titu1994** [comment]
Dunno what "validate" is but basically your fine-tune data has characters that did not exist in the pretrained models vocab set, so during training it malled to ?? And the model learned to predict that. 

You need to either change your Tokenizer and then train on your data or preprocess your data such that there are no ?? after doing a detokenize(tokenize(text)) pass. 

During training and validation - the ground truth text will go through above, and so even the ground truth will be corrupted with ??, which then matches the results of your models predictions with ?? so the results look fine. But model.transcribe() does not put your ground truth through this tokenize-detokenize step so the issue becomes apparent.

**Comment by carlosfranzreb** [comment]
Thanks for the quick reply!

I am using German Common Voice for fine-tuning, which was also used to train the model. So the characters should be fine, right? Also, three of the four characters replaced with `??` in the example above are all capital letters that also exist in the English vocabulary (N, E and M). The last one stands for nothing, as the sentence as already ended. Maybe it refers to the end of sentence.

The tokenizer should be fine. Can you think of anything else? Thanks again for your help.

**Comment by titu1994** [comment]
We normalize all ASR datasets, rather than keep all tokens of the MCV and other datasets, so not necessary that the token map is exactly the same.

**Comment by carlosfranzreb** [comment]
The transcription works properly on the same MCV subset on the model you provide. It stops working once I fine-tune the model and load the checkpoint. So the problem cannot be related to the tokens.

Both the model you provide and the same model after fine-tuning have the same values for the following attributes:

- `model_path`: `/tmp/.../4191e9ad5a794698a366c376cd410d89_tokenizer.model`
- `tokenizer_dir`: `/tokenizers/tokenizer_spe_unigram_v1024/`
- `tokenizer_type`: `bpe`
- `vocab_path`: `/tmp/.../88ae13f5c6fa49579cae0305da764a6c_vocab.txt`
- `spe_vocab_path`: `/tmp/.../8ae19b0527cb432eac7b14e40bc1dc30_tokenizer.vocab`

They have different values for the attribute `model_guid`, but I assume that is a unique identifier for each experiment.

**Comment by carlosfranzreb** [comment]
Both `transcribe` and `validation_step` run the model's encoder. Afterwards, their behavior differs:

- `transcribe` calls the method `rnnt_decoder_predictions_tensor` from a `AbstractRNNTDecoding` object.
- `validation_step` runs the joint model without running the decoder first (`compute_eval_loss` is set to false). The joint model also calls the method `rnnt_decoder_predictions_tensor` after some other steps.

**Comment by titu1994** [comment]
The Tokenizer does not support upper case characters - we always normalize to lower case. To test it, simply load the original model, and use model.tokenizer.tokenizer.text_to_ids("some str") and then call ...ids_to_text(prev_output). Any upper case character will nap to 0 - ??

**Comment by carlosfranzreb** [comment]
When I try to do that, both with the model you provide as is and the fine-tuned version, an error is raised:

```
'SentencePieceProcessor' object has no attribute 'text_to_ids'
```

**Comment by titu1994** [comment]
Oh ok it has different method names - you should be able to drop the second tokenizer in that command, the Nemo wrapper should work with those methods

**Comment by carlosfranzreb** [comment]
Alright, now both models output the text I wrote at the beginning, with all capital letters replaced by `??`. But when I run the model without fine-tuning, the output of the `transcribe` does not contain is written in lower-case, while the output of the fine-tuned model contains these `??` characters. Why is that? The `transcribe` method does not receive the ground truth, so why does it try to output capital letters? And it only tries that after fine-tuning.

Where should I look for a difference between the two models?

**Comment by carlosfranzreb** [comment]
I have two models:

- baseline: `EncDecRNNTBPEModel.from_pretrained(stt_de_conformer_transducer_large)`
- finetuned: Initialized as baseline, but the state dict is updated with the new weights.

The dataloaders, initialized with `EncDecRNNTBPEModel._setup_dataloader_from_config` of both models are the same.

The `transcribe` method, which I am having trouble with, consists running the encoder (called with the forward pass) and the decoder and joint models, called with `model.decoding.rnnt_decoder_predictions_tensor`.

These are the results when combining each encoder with each decoder+joint:

- baseline encoder + baseline decoder+joint: `nun müssen meines erachtens konkrete und erkennbare maßnahmen ergriffen werden` (0 errors)
- baseline encoder + finetuned decoder+joint: `nun müssen meines erachtens konkrete und erkennbare  ⁇ aßnahmen ergriffen werden` (1 error: predicted `??` instead of `m`).
- finetuned encoder + baseline decoder+joint: `urs meines erachtens konkrete und erkennbare nahmen ergriffen werden` (3 errors)
- finetuned encoder + finetuned decoder+joint: ` ⁇ un müssen meines  ⁇ rachtens konkrete und erkennbare  ⁇ aßnahmen ergriffen werden ⁇ ` (as many errors as there are `??`: all other characters are correct)

The finetuned decoder+joint is the only one that outputs `??`. The baseline decoder+joint never outputs that symbol. Why would a decoder+joint output that?

The output of finetuned encoder + finetuned decoder+joint is correct besides the `??`. During training, the WER of the model improved. I doubt the improvement was because of predicting `??`. And I don't understand how this is allowed. This has nothing to do with the ground truth containing capital letters, as the ground truth is not fed to the `transcribe` method.

**Comment by titu1994** [comment]
> The finetuned decoder+joint is the only one that outputs ??. The baseline decoder+joint never outputs that symbol. Why would a decoder+joint output that?

As I've mentioned before - before fine tuning, your model gets audio and emits only lower case text. 

During finetuning and validate - your ground truth which has capital letters - goes through detok(tok(text)) step - and now all capital letters are ?? Due to the tokenizer. Therefore the model learns to predict ?? For all capital letters because that is the ground truth shown to it. 

You also should not randomly combine baseline encoder with finetuned decoder or vice versa unless you froze your encoder (and even then I would not advise if).

**Comment by titu1994** [comment]
> I doubt the improvement was because of predicting ??.

This improvement is faulty - both the ground truth and the models predictions are with ?? - during training and validate

Only model transcribe () will have capital ground truth and models prediction with ?? For finetuned model.

**Comment by carlosfranzreb** [comment]
Alright, now I understand your point. I will run the fine-tuning again with all the texts lower-cased. Thank you for your help and patience!
